sequence_processing_pipeline/Pipeline.py

```diff
@@ -365,15 +396,35 @@ def _validate_mapping_file(mapping_file_path):
                'target_subfragment', 'primer', 'primer_plate', 'sample_name',
                'run_center', 'primer_date', 'target_gene', 'processing_robot',
                'extractionkit_lot'}
+
         try:
             df = pd.read_csv(mapping_file_path, delimiter='\t')
 
-            # if the two sets of headers are equal, then return the dataframe
-            # and no error message.
-            if len(exp - set(df.columns)) == 0:
-                return df
+            # rename all columns to their lower-case versions.
+            # we will want to return this version to the user.
+            df.columns = df.columns.str.lower()
 
-            msg = 'missing columns: %s' % ', '.join(exp - set(df.columns))
+            # if the two sets of headers are equal, verify there are no
+            # duplicate sample-names and then return the dataframe and no
+            # error message.
+            if len(exp - set(df.columns)) == 0:
+                sample_names = df['sample_name'].tolist()
+                dupes = Counter(sample_names)
+                dupes = [x for x in dupes if dupes[x] > 1]
+                if dupes:
+                    msg = ('duplicate sample-names detected: '
+                           '%s' % ', '.join(dupes))
+                else:
+                    return df
+            else:
+                if exp - set(df.columns) == exp:
+                    # since pandas.read_csv() will successfully open a sample-
+                    # sheet, if none of the expected columns exist then assume
+                    # that this file is not a mapping file.
+                    msg = "not a mapping-file"
+                else:
+                    msg = 'missing columns: %s' % ', '.join(
+                        exp - set(df.columns))
```

Comment:
Just noticed that this has an "interesting" logic, basically you `return df` in the middle of the function but you keep the msg and raise an error at the end. I normally have seen the opposite: you raise an error in the middle of the code and `return df` at the end (if no errors where found). The only time I have seen this, that I can remember, is when you can have multiple messages in a single run and you initialize the buffer `msgs = []` and on each new message you append to finally have a `if msgs: raise XX else: return XXX`, which is not the case here.

Reply:
I think it's a clean way to capture external errors like ParserError as well as raise our own specific errors.

Reply:
Ty though, I will keep your comments in mind for the future.